File: auth-service/src/main/java/com/ilogos/auth_service/controller/ExceptionController.java
package com.ilogos.auth_service.controller;

import java.util.ArrayList;
import java.util.List;

import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

import com.ilogos.auth_service.exceptions.ExceptionWithStatus;
import com.ilogos.auth_service.model.response.ErrorResponse;

import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.security.SignatureException;
import jakarta.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@ControllerAdvice
public class ExceptionController {
    @ExceptionHandler(ExceptionWithStatus.class)
    public ResponseEntity<ErrorResponse> handleExceptionWithStatus(ExceptionWithStatus ex) {
        return ErrorResponse.response(ex.getStatus(), ex.getMessage());
    }

    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<ErrorResponse> handleConstraintViolationException(ConstraintViolationException ex) {
        List<String> errors = ex.getConstraintViolations().stream()
                .map(violation -> violation.getPropertyPath() + ": " + violation.getMessage())
                .toList();

        return ErrorResponse.response(HttpStatus.BAD_REQUEST, errors);
    }

    @ExceptionHandler(DataIntegrityViolationException.class)
    public ResponseEntity<ErrorResponse> handleDataIntegrityViolationException(DataIntegrityViolationException ex) {
        String msg = ex.getMostSpecificCause().getMessage();
        if (msg != null) {
            if (msg.contains("email")) {
                return ErrorResponse.response(HttpStatus.BAD_REQUEST, "Email already used");
            } else if (msg.contains("username")) {
                return ErrorResponse.response(HttpStatus.BAD_REQUEST, "Username already used");
            }
        }

        return ErrorResponse.response(HttpStatus.BAD_REQUEST, ex);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleException(Exception ex) {
        return ErrorResponse.response(HttpStatus.INTERNAL_SERVER_ERROR, ex);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleValidationExceptions(MethodArgumentNotValidException ex) {
        List<String> errors = new ArrayList<>();
        ex.getBindingResult().getFieldErrors()
                .forEach(error -> errors.add("%s: %s".formatted(error.getField(), error.getDefaultMessage())));
        return ErrorResponse.response(HttpStatus.BAD_REQUEST, errors);
    }

    @ExceptionHandler(SignatureException.class)
    public ResponseEntity<ErrorResponse> handleException(SignatureException ex) {
        return ErrorResponse.response(HttpStatus.BAD_REQUEST, ex);
    }

    @ExceptionHandler(MalformedJwtException.class)
    public ResponseEntity<ErrorResponse> handleMalformedJwtException(MalformedJwtException ex) {
        return ErrorResponse.response(HttpStatus.BAD_REQUEST, "Malformed JWT");
    }

}

File: course-service/src/main/kotlin/com/ilogos/course/exception/ExceptionController.kt
package com.ilogos.course.exception

import com.ilogos.course.response.ErrorResponse
import io.jsonwebtoken.MalformedJwtException
import io.jsonwebtoken.security.SignatureException
import jakarta.validation.ConstraintViolationException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class ExceptionController {

    private val log = LoggerFactory.getLogger(ExceptionController::class.java)

    @ExceptionHandler(ExceptionWithStatus::class)
    fun handleExceptionWithStatus(ex: ExceptionWithStatus) =
        ErrorResponse.response(ex.status, ex.message)

    @ExceptionHandler(ConstraintViolationException::class)
    fun handleConstraintViolationException(
        ex: ConstraintViolationException
    ): ResponseEntity<ErrorResponse> {
        val errors = ex.constraintViolations.map { "${it.propertyPath}: ${it.message}" }
        return ErrorResponse.response(HttpStatus.BAD_REQUEST, errors)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationExceptions(
        ex: MethodArgumentNotValidException
    ): ResponseEntity<ErrorResponse> {
        val errors = ex.bindingResult.fieldErrors.map { "${it.field}: ${it.defaultMessage}" }
        return ErrorResponse.response(HttpStatus.BAD_REQUEST, errors)
    }

    @ExceptionHandler(SignatureException::class)
    fun handleSignatureException(ex: SignatureException) =
        ErrorResponse.response(HttpStatus.BAD_REQUEST, ex)

    @ExceptionHandler(MalformedJwtException::class)
    fun handleMalformedJwtException() =
        ErrorResponse.response(HttpStatus.BAD_REQUEST, "Malformed JWT")

    @ExceptionHandler(Exception::class)
    fun handleGenericException(ex: Exception): ResponseEntity<ErrorResponse> {
        log.error("Unexpected error occurred", ex)
        return ErrorResponse.response(HttpStatus.INTERNAL_SERVER_ERROR, ex)
    }
}
